docs: Update documentation on CREATE ROLE WITH HASHED PASSWORD

dawmd opened the pull request:
As part of #18750, we added a CQL statement CREATE ROLE WITH SALTED HASH that prevented hashing a password when creating a role, effectively leading to inserting a hash given by the user directly into the database. In #21350, we noticed that Cassandra had implemented a CQL statement of similar semantics but different syntax. We decided to rename Scylla's statement to be compatible with Cassandra. Unfortunately, we didn't notice one more difference between what we had in Scylla and what was part of Cassandra.

Scylla's statement was originally supposed to only be used when restoring the schema and the user needn't have to be aware of its existence at all: the database produced a sequence of CQL statements that the user saved to a file and when a need to restore the schema arose, they would execute the contents of the file. That's why that although we documented the feature, it was only done in the necessary places. Those that weren't related to the backup & restore procedure were deliberately skipped.

Cassandra, on the other hand, added the statement for a different purpose (for details, see the relevant issue) and it was supposed to be used by the user by design. The statement is also documented as such.

Since we want to preserve compatibility with Cassandra, we document the statement and its semantics in the user documentation, explicitly implying that it can be used by the user.

We also add a test verifying that logging in works correctly.

Fixes scylladb/scylladb#21691

Backport: not needed. The relevant code didn't make it to 6.2 or any previous version of OSS.

piodul commented on docs/operating-scylla/security/authorization.rst:
```diff
@@ -91,6 +91,30 @@ If the option is used and the role exists, the statement is a no-op::
+
+The provided hashed password must be an encrypted string using the library function `crypt(3)` specified by POSIX,
+and it must use one of the encryption algorithms that ScyllaDB supports: bcrypt, sha512crypt, sha256crypt, md5crypt.
+Not abiding by those requirements may result in undefined behavior, so you should be very careful when executing the statement.
```
"Undefined behavior" sounds very scary. What's the worst that can actually happen? The user won't be able to log in, or the whole database crashes?

If it's the latter then we should improve how we handle malformed hashed password.

dawmd replied:
What happens now is we insert the hashed password into the database no matter what it is, but if it's invalid (i.e. it doesn't follow one of those algorithms), trying to log in will result in a system error—that's what I've observed. In that case, the best the user can do is remove the role from the database and create it again. I used the wording with undefined behaviour because there might be more side effects of that that I haven't thought of.

piodul commented on docs/operating-scylla/security/authorization.rst:
```diff
@@ -91,6 +91,30 @@ If the option is used and the role exists, the statement is a no-op::
+the database directly, with no additional processing of it. The statement only works with an authenticator that
+uses passwords, e.g. PasswordAuthenticator. It can only be executed by a role with the superuser privilege.
+
+The provided hashed password must be an encrypted string using the library function `crypt(3)` specified by POSIX,
+and it must use one of the encryption algorithms that ScyllaDB supports: bcrypt, sha512crypt, sha256crypt, md5crypt.
```
Would it be possible to provide an example on how to produce such a hash?

dawmd replied:
I could try to add a block of C code with comments showcasing how to use `crypt(3)`. It would be way more difficult to explain it in words alone, though.

dawmd replied:
Note: in Scylla, we actually use `crypt_r(3)`, which is an alternative for `crypt(3)`, but as far as I can tell the only difference between the functions is `crypt_r(3)` is re-entrant (for that, it takes an additional parameter that it operates on instead of using a static buffer).

dawmd commented:
Version 2:
* Got rid of the undefined behaviour wording in the documentation. Added examples of how to generate a hashed password using `crypt(3)` and how to create a role with the statement.
* Added a test verifying that logging in works correctly with all of the supported encryption formats.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13900/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13900/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/auth_test
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2352/)
:x: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/150/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/748/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5473/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13487/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13900/)
#### Failed Tests (3/36271):
- [test_mv_topology_change.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13900/testReport/junit/topology_custom/test_mv_topology_change/Tests___Unit_Tests___test_mv_topology_change_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_topology_change.dev.2)
- [test_topology_ops[False].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13900/testReport/junit/topology_experimental_raft/test_topology_ops/Tests___Unit_Tests___test_topology_ops_False__debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_ops[False].debug.3)
- [cqlpy.cassandra_tests/validation/entities/static_columns_test.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13900/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_static_columns_test_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/static_columns_test.debug.2)
#### Build Details:
- Duration: 8 hr 18 min
- Builder: i-0912d416b2a4a2a8e (m5ad.8xlarge)

dawmd commented:
>     * [test_mv_topology_change.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13900/testReport/junit/topology_custom/test_mv_topology_change/Tests___Unit_Tests___test_mv_topology_change_dev_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_topology_change.dev.2)

scylladb/scylladb#21912

>     * [test_topology_ops[False].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13900/testReport/junit/topology_experimental_raft/test_topology_ops/Tests___Unit_Tests___test_topology_ops_False__debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_ops%5BFalse%5D.debug.3)

scylladb/scylladb#21949

>     * [cqlpy.cassandra_tests/validation/entities/static_columns_test.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13900/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cqlpy_cassandra_tests_validation_entities_static_columns_test_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cqlpy.cassandra_tests/validation/entities/static_columns_test.debug.2)

scylladb/scylladb#21925

Rerunning CI

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13958/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13958/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/auth_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/795/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2404/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5520/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13534/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/197/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13958/)
#### Failed Tests (3/36489):
- [test_simple_removenode_5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13958/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_5)
- [test_simple_removenode_5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13958/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_5_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_5)
- [test_replace_with_no_normal_token_owners_in_dc[False].release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13958/testReport/junit/topology_custom/test_tablets/Tests___Unit_Tests___test_replace_with_no_normal_token_owners_in_dc_False__release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_no_normal_token_owners_in_dc[False].release.1)
#### Build Details:
- Duration: 4 hr 10 min
- Builder: spider4.cloudius-systems.com

dawmd commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13958/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13958/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/auth_test ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/795/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2404/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5520/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13534/) ✅ - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/197/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13958/)
> #### Failed Tests (3/36489):
> 
>     * [test_simple_removenode_5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13958/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_5) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_5)
> 
>     * [test_simple_removenode_5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13958/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_5_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_5)

scylladb/scylladb#20814

> 
>     * [test_replace_with_no_normal_token_owners_in_dc[False].release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13958/testReport/junit/topology_custom/test_tablets/Tests___Unit_Tests___test_replace_with_no_normal_token_owners_in_dc_False__release_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_no_normal_token_owners_in_dc%5BFalse%5D.release.1)

scylladb/scylladb#22229

> 
> 
> #### Build Details:
> 
>     * Duration: 4 hr 10 min
> 
>     * Builder: spider4.cloudius-systems.com

Rerunning CI

nyh commented:
Thanks, looks very good.
Because this PR doesn't change any Scylla code, any CI failures (except if the new test introduced in the PR) is irrelevant, so the previous CI which failed only on irrelevant stuff is good enough for me (however, thanks for looking into the flaky tests anyway!). I'll merge now.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14437/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14437/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/auth_test
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2779/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13909/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5899/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14437/)
#### Failed Tests (2/37070):
- [test_view_build_status_cleanup_on_remove_node.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14437/testReport/junit/topology_custom/test_view_build_status/Tests___Unit_Tests___test_view_build_status_cleanup_on_remove_node_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_view_build_status_cleanup_on_remove_node.release.1)
- [test_view_build_status_cleanup_on_remove_node.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14437/testReport/junit/topology_custom/test_view_build_status/Tests___Unit_Tests___test_view_build_status_cleanup_on_remove_node_release_1_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_view_build_status_cleanup_on_remove_node.release.1)
#### Build Details:
- Duration: 11 hr
- Builder: i-0128492b9a329ca3a (m5ad.8xlarge)
